sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala

```diff
@@ -187,6 +197,41 @@ object RapidsBufferCatalog extends Logging {
       initialSpillPriority: Long): Unit =
     deviceStorage.addBuffer(id, buffer, tableMeta, initialSpillPriority)
 
+  /**
+   * Adds a batch to the device storage, taking ownership of the batch. It will try to be smart
+   * and only copy the data into a contiguous buffer if the incoming batch is not already
+   * contiguous.
+   * @param id buffer ID to associate with this buffer
+   * @param batch batch to be processed
+   * @param initialSpillPriority starting spill priority value for the buffer
+   */
+  def addBatch(
+      id: RapidsBufferId,
+      batch: ColumnarBatch,
+      initialSpillPriority: Long): Unit = {
+    val numColumns = batch.numCols()
+    if (numColumns > 0 && batch.column(0).isInstanceOf[GpuCompressedColumnVector]) {
+      val cv = batch.column(0).asInstanceOf[GpuCompressedColumnVector]
+      deviceStorage.addBuffer(id, cv.getBuffer, cv.getTableMeta, initialSpillPriority)
+    } else if (numColumns > 0 &&
+        (0 until numColumns)
+            .forall(i => batch.column(i).isInstanceOf[GpuColumnVectorFromBuffer])) {
+      val cv = batch.column(0).asInstanceOf[GpuColumnVectorFromBuffer]
+      withResource(GpuColumnVector.from(batch)) { table =>
+        deviceStorage.addTable(id, table, cv.getBuffer, initialSpillPriority)
+      }
+    } else {
+      withResource(batch) { batch =>
+        withResource(GpuColumnVector.from(batch)) { tmpTable =>
+          val contigTables = tmpTable.contiguousSplit(batch.numRows())
+          val tab = contigTables.head
+          contigTables.tail.safeClose()
+          deviceStorage.addTable(id, tab.getTable, tab.getBuffer, initialSpillPriority)
```

Comment:
What about the case where the batch has zero columns and only rows?  We should not try to contiguous split such batches nor add them to the spillable cache since there's no data to be spilled.

Reply:
If you really want me to try and handle that case separately here I can, but it will blow up when we try to turn it into a table.  Do you want me to throw a clearer error message in those cases?  Should I just update the docs?  Not sure how you want me to fix this.

Reply:
We can't let it blow up, right?  ColumnarBatch with only rows and no columns is a legitimate batch that happens in real queries, so unless that case is being handled somewhere else we need to handle it here.  Apologies if I'm missing something.

Reply:
> it will blow up when we try to turn it into a table.

Yes, that's true, which makes me wonder how we're handling these row-only batches today as they pass through the coalesce iterator.  This must be handled at a higher layer, preventing us from ever trying to call a cudf coalesce with such batches.

Reply:
It does not happen right now.  If it did there are all kinds of other places where the code would blow up with the same error, just later on in the processing.  If you want me to try and handle it in a generic way I can do that, but I would probably put a warning on this API about it and update SpillableColumnarBatch  to detect it and handle it separately.

Reply:
> It does not happen right now.

Does the plugin avoid placing coalesce batch in the plan if the data schema at that point is just row counts?  I'm not quite sure how we're avoiding trying to coalesce row-count-only batches today.

Agree that if we're sure this isn't happening then it's not something we need to worry about as part of this PR.  If it does happen it will explode, although probably not in an obvious way.

Reply:
Actually I think I'll just move this into SpillableColumnarBatch so it is all hidden away and does the right thing all the time.